src/cleanlab_tlm/internal/rag.py

```diff
@@ -0,0 +1,121 @@
+
+    When tool call handling is enabled and a tool call is detected:
+    - Filters out evals that have response_identifier (these would get None scores)
+    - Calls the original method with filtered evals via a context wrapper
+    - Adds None scores for the filtered evals
+    - Returns the combined result
+
+    This implementation avoids modifying the original instance state to prevent
+    race conditions in concurrent async operations.
+    """
+
+    @wraps(func)
+    async def wrapper(self: Any, **kwargs: Any) -> ResponseT:
+        response = kwargs.get("response", {})
+        response_text = response.get("response", "")
+        is_tool_call = _is_tool_call_response(str(response_text))
+
+        # If not a tool call, just call the original method
+        if not is_tool_call:
+            return await func(self, **kwargs)
+
+        # It's a tool call - determine which evals to process vs. filter
+        # Default behavior:
+        #   - Evals with response_identifier are filtered out (score None)
+        #   - Evals without response_identifier are still evaluated normally
+        # Optional per-eval overrides via instance-level include/exclude name sets:
+        #   - If name in exclude set, filter (score None)
+
+        exclude_names = set(getattr(self, "_tool_call_eval_exclude_names", set()) or set())
+
+        evals_to_process = []
+        tool_call_filtered_evals = []
+
+        for eval_obj in self._evals:
+            # Start from default filtering decision
+            is_filtered = eval_obj.response_identifier is not None and eval_obj.name in exclude_names
+
+            # Apply exclude override (forces filtering)
+            if eval_obj.name in exclude_names:
+                is_filtered = True
+
+            if is_filtered:
+                tool_call_filtered_evals.append(eval_obj)
+            else:
+                evals_to_process.append(eval_obj)
+
+        # Create a context wrapper that temporarily provides filtered evals
+        # without modifying the original instance
+        class _EvalsContextWrapper:
+            def __init__(self, original_instance: Any, filtered_evals: list[Any]):
+                self._original = original_instance
+                self._filtered_evals = filtered_evals
+
+            def __getattr__(self, name: str) -> Any:
+                if name == "_evals":
+                    return self._filtered_evals
+                return getattr(self._original, name)
+
+            def __repr__(self) -> str:
+                return repr(self._original)
+
+            def __str__(self) -> str:
+                return str(self._original)
+
+        # Use the wrapper instance to call the original method
+        wrapper_instance = _EvalsContextWrapper(self, evals_to_process)
+        backend_response = await func(wrapper_instance, **kwargs)
+
+        # Add None scores for the filtered evals
+        for eval_obj in tool_call_filtered_evals:
```

Comment:
This will make it such that the [order of the keys](https://github.com/cleanlab/cleanlab-tlm/blob/main/src/cleanlab_tlm/internal/api/api.py#L526-L529) might not be adhered -- is there a way to keep that original order?